File: #105_Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
import java.util.HashMap;

public class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
    	if (preorder.length == 0 || inorder.length == 0) return null;
    	boolean[] creat = new boolean[inorder.length];
    	HashMap<Integer, Integer> map = new HashMap<Integer, Integer>();
    	/* number -> index */
        for (int i = 0; i < inorder.length; i++) map.put(inorder[i], i);
        HashMap<Integer, TreeNode> tree = new HashMap<Integer, TreeNode>();
        TreeNode root = new TreeNode(preorder[0]);
        /* number -> treenode */
        tree.put(preorder[0], root);
	    creat[map.get(preorder[0])] = true;
        for (int i = 1; i < preorder.length; i++) {
        	int index = map.get(preorder[i]);
        	System.out.println("index:" + index);
        	/* look into its right side */
        	for (int j = index + 1; j < inorder.length; j++) {
        		if (creat[j]) {
        			// System.out.println("Find");
        			TreeNode p = tree.get(inorder[j]);
        			if (p.left == null) {
        				p.left = new TreeNode(preorder[i]);
        				tree.put(preorder[i], p.left);
        				creat[index] = true;
        				break;
        			}
        		}
        	}
        	if (!creat[index]) {
	        	/* look into its left side */
	        	for (int j = index - 1; j >= 0; j--) {
	        		if (creat[j]) {
	        			System.out.println("Find");
	        			TreeNode p = tree.get(inorder[j]);
	        			p.right = new TreeNode(preorder[i]);
	        			tree.put(preorder[i], p.right);
	        			creat[index] = true;
	        			break;
	        		}
	        	}	
	        }
        }
        return root;
    }
    public static void main(String[] args) {
    	Solution sol = new Solution();
    	int[] pre = {1, 2};
    	int[] in = {1, 2};
    	TreeNode r = sol.buildTree(pre, in);
    	// r.inorder();
    }
}

File: #105_Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal/Solution.py
class TreeNode:
	def __init__(self, x):
		self.val = x
		self.left = None
		self.right = None

class Solution:

	def __init__(self):
		self.map = None

	# @param {integer[]} preorder
	# @param {integer[]} inorder
	# @return {TreeNode}
	def buildTree(self, preorder, inorder):
		if not preorder or not inorder:
			return None
		self.map = {}
		for index, value in enumerate(inorder):
			self.map[value] = index
		return self.build(preorder, inorder, 0, len(inorder) - 1)

	def build(self, preorder, inorder, start, end):
		if not preorder or start > end:
			return None
		root = TreeNode(preorder[0])
		index = self.map[preorder[0]]
		leftElements = index - start
		root.left = self.build(preorder[1:1 + leftElements], inorder, start, index - 1)
		root.right = self.build(preorder[1 + leftElements:], inorder, index + 1, end)
		return root